containers/unit_tests/TestUnorderedSet.hpp

```diff
@@ -0,0 +1,122 @@
+
+//! @test Ensure that inserting, erasing and rehasing an unordered set works.
+TEST(TEST_CATEGORY, UnorderedSet_insert_erase_and_rehash) {
+  /**
+   * The set will be inserted for @ref size_all entries,
+   * and then @ref size_erased will be erased from it before rehash.
+   * Once rehasing is done, the capacity of the set is expected to
+   * be smaller than the one of the initial set.
+   */
+  constexpr size_t size_all    = 2579;
+  constexpr size_t size_erased = 568;
+  static_assert(size_all > size_erased);
+
+  using set_type = Kokkos::UnorderedMap<size_t, void, TEST_EXECSPACE>;
+  static_assert(set_type::is_set);
+
+  using policy_type = Kokkos::RangePolicy<TEST_EXECSPACE>;
+  using index_type  = typename policy_type::index_type;
+
+  using index_view_type = Kokkos::View<index_type[size_erased], TEST_EXECSPACE>;
+
+  //! Ensure that the space instance is not the default one.
+  const auto space = Kokkos::Experimental::partition_space(TEST_EXECSPACE{}, 1)[0];
+
+  //! Initialize the set.
+  set_type uset(Kokkos::view_alloc(space, "test uset"), size_all);
+  const auto initial_capacity = uset.capacity();
+
+  ASSERT_GE(uset.capacity(), size_all);
+  ASSERT_EQ(uset.size()    , 0);
+
+  //! Insert @ref size_all keys.
+  Kokkos::parallel_for(
+    "uset test - insert " + std::to_string(size_all) + " values",
+    policy_type(space, 0, size_all),
+    KOKKOS_LAMBDA(const index_type index) {
+        Kokkos::UnorderedMapInsertResult res = uset.insert(index);
+        if( ! res.success() ) Kokkos::abort("Cannot insert index.");
+    }
+  );
+
+  space.fence();
+
+  ASSERT_GE(uset.capacity(), size_all);
+  ASSERT_EQ(uset.size()    , size_all);
+
+  //! Always try to erase truly random indices by initializing the random pool randomly.
+  std::random_device dev;
+  std::mt19937 rng(dev());
+  std::uniform_int_distribution<int> dist(0, std::numeric_limits<int>::max());
+
+  //! Generate random indices between @c 0 and @ref size_all. Those will be erased.
+  Kokkos::Random_XorShift64_Pool<TEST_EXECSPACE> generator(dist(rng));
+  const index_view_type indices_erased(Kokkos::view_alloc(Kokkos::WithoutInitializing, space, "indices that will be erased"));
+  Kokkos::fill_random(space, indices_erased, generator, size_all);
+  Kokkos::sort(space, indices_erased);
+
+  //! Filter out any duplicated index.
+  namespace KE = Kokkos::Experimental;
+  const auto indices_erased_end  = KE::unique(space, KE::begin(indices_erased), KE::end(indices_erased));
+  const auto indices_erased_size = KE::distance(KE::begin(indices_erased), indices_erased_end);
+  const auto expected_uset_size  = size_all - indices_erased_size;
+
+  //! Debug information.
+  std::cout << "> initial uset   size: " << size_all            << std::endl;
+  std::cout << "> initial erased size: " << size_erased         << std::endl;
+  std::cout << "> final   erased size: " << indices_erased_size << std::endl;
+  std::cout << "> final   used   size: " << expected_uset_size  << std::endl;
```

Comment:
Let's not clutter the test output with debug output.
```suggestion
```

Reply:
Ok, removed :heavy_check_mark: